File: HeatTransferExperiment.py
def experement(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th):

    # Определение шагов сетки
    h = R / (N - 1)
    tau = t_end / 100.0

    # Инициализация полей температуры
    T = [T0] * N
    alfa = [0.0] * N
    beta = [0.0] * N

    # Интегрирование уравнения теплопроводности
    time = 0
    while time < t_end:
        time += tau
        alfa[0] = 1.0
        beta[0] = 0.0

        for i in range(1, N - 1):
            ai = 0.5 * lamda * (2 * i + 1) / (h**2 * i)
            ci = 0.5 * lamda * (2 * i - 1) / (h**2 * i)
            bi = ai + ci + ro * c / tau
            fi = -ro * c * T[i] / tau
            alfa[i] = ai / (bi - ci * alfa[i - 1])
            beta[i] = (ci * beta[i - 1] - fi) / (bi - ci * alfa[i - 1])

        T[N - 1] = Th
        for i in range(N - 2, -1, -1):
            T[i] = alfa[i] * T[i + 1] + beta[i]
            if alfa[i] == 0*T[i + 1] == -beta[i]:
                print(alfa[i], T[i + 1], beta[i])

    # Сохранение результатов в файлы
    with open(f'{name_of_experement}_res.txt', 'w') as f:
        f.write(f'Радиус цилиндра R = {R:.4f}\n')
        # Остальные параметры здесь...

    with open(f'{name_of_experement}_tempr.txt', 'w') as g:
        for i in range(N):
            g.write(f'{h * i:.3f} {T[i]:.5f}\n')

    return [h * i for i in range(N)], T

flag = input('Введите режим работы программы текстом 1) test 2) keyboard 3) file: ')
data_list = []
if flag == 'test':
    # Пример использования:
    # Ввод параметров
    name_of_experement = 'test'
    N = 100
    R = 0.1
    lamda = 0.12
    ro = 400
    c = 500
    T0 = 20
    Th = 50
    t_end = 3600
    d = experement(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th)
elif flag == 'keyboard':
    name_of_experement = input('Введите название эксперимента: ')
    N = int(input('Введите количество пространственных узлов, N: '))
    R = float(input('Введите радиус цилиндра, R: '))
    lamda = float(input('Введите коэффициент теплопроводности материала пластины, lamda: '))
    ro = float(input('Введите плотность материала пластины, ro: '))
    c = float(input('Введите теплоемкость материала пластины, c: '))
    T0 = float(input('Введите начальную температуру, T0: '))
    Th = float(input('Введите температуру на границе r=R, Th: '))
    t_end = float(input('Введите окончание по времени, t_end: '))
    d = experement(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th)
elif flag == 'file':
    input_file_name = input('Введите название файла для считывания исходных данных: ')
    with open(f'{input_file_name}.txt', 'r') as f:
        name_of_experement = f.readline().split()[-1]
        N = int(f.readline().split()[-1])
        R = float(f.readline().split()[-1])
        lamda = float(f.readline().split()[-1])
        ro = float(f.readline().split()[-1])
        c = float(f.readline().split()[-1])
        T0 = float(f.readline().split()[-1])
        Th = float(f.readline().split()[-1])
        t_end = float(f.readline().split()[-1])
        d = experement(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th)

File: HeatTransferExperiment.pas
program HeatTransferExperiment;

const
    mf = 100;

type
    vector = array[1..mf] of real;

var
    N: integer;
    T, alfa, beta: vector;
    ai, bi, ci, fi: real;
    lamda, ro, c, h, tau, R, T0, Th, t_end, time: real;
    f, g, input_file: text;
    name_of_experement, flag, input_file_name: string;

procedure experiment(name_of_experement: string; N: integer; t_end: real; R: real; lamda: real; ro: real; c: real; T0: real; Th: real);
begin
    h := R / (N - 1);
    tau := t_end / 100.0;

    for var i := 1 to N do
        T[i] := T0;

    time := 0;
    while time < t_end do
    begin
        time := time + tau;
        alfa[1] := 1.0;
        beta[1] := 0.0;

        for var i := 2 to N - 1 do
        begin
            ai := 0.5 * lamda * (2 * i + 1) / (sqr(h) * i);
            ci := 0.5 * lamda * (2 * i - 1) / (sqr(h) * i);
            bi := ai + ci + ro * c / tau;
            fi := -ro * c * T[i] / tau;
            alfa[i] := ai / (bi - ci * alfa[i - 1]);
            beta[i] := (ci * beta[i - 1] - fi) / (bi - ci * alfa[i - 1]);
        end;

        T[N] := Th;
        for var i := N - 1 downto 1 do
            T[i] := alfa[i] * T[i + 1] + beta[i];
    end;

    Assign(f, name_of_experement + '_res.txt');
    Rewrite(f);
    Writeln(f, 'name_of_experement = ', name_of_experement);
    Writeln(f, 'Радиус цилиндра N = ', N);
    Writeln(f, 'Радиус цилиндра t_end = ', t_end);
    Writeln(f, 'Радиус цилиндра R = ', R:6:4);
    Writeln(f, 'Радиус цилиндра lamda = ', lamda:6:4);
    Writeln(f, 'Радиус цилиндра ro = ', ro:6:4);
    Writeln(f, 'Радиус цилиндра c = ', c:6:4);
    Writeln(f, 'Радиус цилиндра T0 = ', T0:6:4);
    Writeln(f, 'Радиус цилиндра Th = ', Th:6:4);

    Close(f);

    Assign(g, name_of_experement + '_tempr.txt');
    Rewrite(g);
    for var i := 1 to N do
        Writeln(g, ' ', h * (i - 1):6:3, ' ', T[i]:8:5);
    Close(g);
end;

begin
    writeln('Введите режим работы программы текстом 1) test 2) keyboard 3) file');
    ReadLn(flag);
    if flag = 'test' then
    begin
      writeln('You choice is test');
      name_of_experement := 'test';
      N := 100;
      R := 0.1;
      lamda := 0.12;
      ro := 400;
      c := 500;
      T0 := 20;
      Th := 50;  
      t_end := 3600;
      experiment(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th);
      end
    else if flag = 'keyboard' then
    begin
      writeln('You choice is keyboard');
      writeln('name_of_experement: ');
      ReadLn(name_of_experement);
      writeln('N: ');
      ReadLn(N); 
      writeln('R: ');
      ReadLn(R); 
      writeln('lamda: ');
      ReadLn(lamda); 
      writeln('ro: ');
      ReadLn(ro); 
      writeln('c: ');
      ReadLn(c); 
      writeln('T0: ');
      ReadLn(T0); 
      writeln('Th: ');
      ReadLn(Th); 
      writeln('t_end: ');
      ReadLn(t_end); 
      experiment(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th);
      end
    else if flag = 'file' then
    begin
      writeln('You choice is file');
      writeln('input_file_name: ');
      ReadLn(input_file_name);
      Assign(input_file, input_file_name + '.txt');
      Reset(input_file);
      Read(input_file, name_of_experement);
      Read(input_file, N);
      Read(input_file, R);
      Read(input_file, lamda);
      Read(input_file, ro);
      Read(input_file, c);
      Read(input_file, T0);
      Read(input_file, Th);
      Read(input_file, t_end);
      input_file.close();
      experiment(name_of_experement, N, t_end, R, lamda, ro, c, T0, Th);
      end
    
end.